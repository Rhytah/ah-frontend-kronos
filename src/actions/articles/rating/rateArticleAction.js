/* eslint-disable func-names */
import axios from 'axios';
import { toast } from 'react-toastify';
import { toastSuccess, toastFailure } from '../../../utils/toast';
import { ArticleConstants } from '../../actionTypes';

const apiUrl = 'https://ah-backend-kronos-staging.herokuapp.com/api/articles/';
const lastPart = '/rate';


const articleRatingAction = (articleSlug, rating) => dispatch => axios.post(`${apiUrl}${articleSlug}${lastPart}`, rating, {
  headers: {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then(() => {
    toast.dismiss();
    toastSuccess('Your rated this article successfully', 'A');
    dispatch({
      type: ArticleConstants.ARTICLE_RATING_SUCCESS,
    });
  })
  .catch((error) => {
    const errorResponse = error.response.data;
    if ({}.hasOwnProperty.call(errorResponse, 'article')) {
      toast.dismiss();
      toastFailure(errorResponse.article.detail, 'A');
      dispatch({
        type: ArticleConstants.ARTICLE_RATING_FAIL,
        payload: errorResponse.article.detail,
      });
    } else {
      dispatch({
        type: ArticleConstants.ARTICLE_RATING_UPDATE,
        payload: errorResponse.errors.error[0],
      });
    }
  });


export default articleRatingAction;
