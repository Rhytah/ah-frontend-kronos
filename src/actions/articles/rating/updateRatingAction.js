/* eslint-disable func-names */
import axios from 'axios';
import { toast } from 'react-toastify';
import { toastSuccess, toastFailure } from '../../../utils/toast';
import { ArticleConstants } from '../../actionTypes';

const apiUrl = 'https://ah-backend-kronos-staging.herokuapp.com/api/articles/';
const lastPart = '/rate';

const articleRatingUpdateAction = (articleSlug, rating) => dispatch => axios.patch(`${apiUrl}${articleSlug}${lastPart}`, rating, {
  headers: {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then((response) => {
    toast.dismiss();
    toastSuccess('Your Rating for Article was Updated!', 'A');
    dispatch({
      type: ArticleConstants.ARTICLE_RATING_UPDATE_SUCCESS,
      payload: response.data.article.rating,
    });
  })
  .catch((error) => {
    toast.dismiss();
    toastFailure('Article rating update failed.', 'A');
    dispatch({
      type: ArticleConstants.ARTICLE_RATING_UPDATE_FAIL,
      payload: error.response,
    });
  });

export default articleRatingUpdateAction;
