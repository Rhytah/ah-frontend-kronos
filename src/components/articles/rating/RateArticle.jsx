/* eslint-disable jsx-a11y/mouse-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import PropTypes from 'prop-types';
import '../../../assets/scss/Landing.scss';

export default function RateArticle({ handleRating, handleMouseOver }) {
  return (
    <div className="ratings-div">
      <span id="star1" onClick={handleRating(1)} onMouseOver={handleMouseOver(1)} className="fa fa-star rating-star rate-star" />
      <span id="star2" onClick={handleRating(2)} onMouseOver={handleMouseOver(2)} className="fa fa-star rating-star rate-star" />
      <span id="star3" onClick={handleRating(3)} onMouseOver={handleMouseOver(3)} className="fa fa-star rating-star rate-star" />
      <span id="star4" onClick={handleRating(4)} onMouseOver={handleMouseOver(4)} className="fa fa-star rating-star rate-star" />
      <span id="star5" onClick={handleRating(5)} onMouseOver={handleMouseOver(5)} className="fa fa-star rating-star rate-star" />
    </div>
  );
}

RateArticle.propTypes = {
  handleRating: PropTypes.func,
  handleMouseOver: PropTypes.func,
};

RateArticle.defaultProps = {
  handleRating: () => {},
  handleMouseOver: () => {},
};
