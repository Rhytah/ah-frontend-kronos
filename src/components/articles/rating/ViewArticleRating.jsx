/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/mouse-events-have-key-events */
import React from 'react';
import { PropTypes } from 'prop-types';
import '../../../assets/scss/Landing.scss';


const isCheckedClass = (position, rating) => {
  let ratingClass = 'fa fa-star rating-star';

  ratingClass = position <= rating ? ratingClass.concat(' checked') : ratingClass;
  return ratingClass;
};
const ViewArticleRating = ({
  rating, handleRating, handleMouseOver, handleMouseOut,
}) => {
  const ratingslist = [1, 2, 3, 4, 5];
  return ratingslist.map((theRating, index) => (
    <span
      // eslint-disable-next-line react/no-array-index-key
      key={index}
      onClick={() => handleRating(theRating)}
      onMouseOver={() => handleMouseOver(theRating)}
      onMouseOut={handleMouseOut}
      className={isCheckedClass(theRating, rating)}
    />
  ));
};

export default ViewArticleRating;

ViewArticleRating.propTypes = {
  rating: PropTypes.number,
  handleRating: PropTypes.func,
  handleMouseOver: PropTypes.func,
  handleMouseOut: PropTypes.func,
};

ViewArticleRating.defaultProps = {
  rating: 0,
  handleRating: () => {},
  handleMouseOver: () => {},
  handleMouseOut: () => {},
};
