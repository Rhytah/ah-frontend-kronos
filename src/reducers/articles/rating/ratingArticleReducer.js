import { ArticleConstants } from '../../../actions/actionTypes';

const rateArticlesReducer = (state = {
  articleRated: null,
  isUpdatingRating: true,
  ratingIsUpdated: null,
  rating: 0,
}, { type, payload }) => {
  switch (type) {
    case ArticleConstants.ARTICLE_RATING_SUCCESS:
      return {
        ...state,
        articleRated: true,
        isUpdatingRating: false,
      };
    case ArticleConstants.ARTICLE_RATING_FAIL:
      return {
        ...state,
        errors: payload,
        articleRated: false,
        isUpdatingRating: false,
      };
    case ArticleConstants.ARTICLE_RATING_UPDATE:
      return {
        ...state,
        errors: payload,
        articleRated: false,
        isUpdatingRating: true,
      };
    case ArticleConstants.ARTICLE_RATING_UPDATE_FAIL:
      return {
        ...state,
        errors: payload,
        ratingIsUpdated: false,
      };
    case ArticleConstants.ARTICLE_RATING_UPDATE_SUCCESS:
      return {
        ...state,
        rating: payload,
        ratingIsUpdated: true,
      };
    default:
      return state;
  }
};

export default rateArticlesReducer;
