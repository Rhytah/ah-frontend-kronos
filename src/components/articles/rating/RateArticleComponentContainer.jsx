/* eslint-disable react/forbid-prop-types */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { toast } from 'react-toastify';
import ViewArticleRating from './ViewArticleRating';
import articleRatingAction from '../../../actions/articles/rating/rateArticleAction';
import articleRatingUpdateAction from '../../../actions/articles/rating/updateRatingAction';
import { toastFailure } from '../../../utils/toast';


export class RateArticleComponentContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      articleSlug: props.articleSlug,
      newRating: 0,
      isRating: false,
    };
  }

  handleRating = async (rating) => {
    const { articleSlug } = this.state;
    const ratingS = {
      rating,
    };
    const { rateArticle, updateArticle } = this.props;

    const isLoggedIn = sessionStorage.getItem('isLoggedIn');

    if (isLoggedIn === null || isLoggedIn === false) {
      toast.dismiss();
      toastFailure('Signin to rate articles.', 'A');
      this.setState({
        isRating: false,
        newRating: 0,
      });
      return;
    }
    rateArticle(articleSlug, ratingS);

    const { ratingsReducer } = this.props;
    const { isUpdatingRating } = ratingsReducer;

    if (isUpdatingRating) {
      await updateArticle(articleSlug, ratingS);
    }
  };

  handleMouseOver = (star) => {
    this.setState({
      isRating: true,
      newRating: star,
    });
  }

  handleMouseOut = () => {
    this.setState({
      isRating: false,
      newRating: 0,
    });
  }

  render() {
    const { isRating, newRating } = this.state;
    if (isRating) {
      return (
        <ViewArticleRating
          rating={newRating}
          handleMouseOut={this.handleMouseOut}
          handleRating={this.handleRating}
          handleMouseOver={this.handleMouseOver}
        />
      );
    }
    const { articleRating } = this.props;
    const { ratingsReducer } = this.props;
    const { rating } = ratingsReducer;

    let theArticleRating;
    if (rating) {
      theArticleRating = parseInt(rating, 10);
    } else {
      theArticleRating = articleRating;
    }

    const loggedIn = JSON.parse(sessionStorage.getItem('isLoggedIn'));

    return (
      loggedIn
        ? (
          <ViewArticleRating
            rating={theArticleRating}
            handleMouseOut={this.handleMouseOut}
            handleRating={this.handleRating}
            handleMouseOver={this.handleMouseOver}
          />
        )
        : <ViewArticleRating rating={theArticleRating} />
    );
  }
}

RateArticleComponentContainer.propTypes = {
  articleSlug: PropTypes.string.isRequired,
  rateArticle: PropTypes.func.isRequired,
  updateArticle: PropTypes.func.isRequired,
  articleRating: PropTypes.number,
  ratingsReducer: PropTypes.object,
};

RateArticleComponentContainer.defaultProps = {
  ratingsReducer: {},
  articleRating: 0,
};

export const mapStateToProps = state => ({
  ratingsReducer: state.rateArticlesReducer,
});


export default connect(
  mapStateToProps,
  {
    rateArticle: articleRatingAction,
    updateArticle: articleRatingUpdateAction,
  },
)(RateArticleComponentContainer);
