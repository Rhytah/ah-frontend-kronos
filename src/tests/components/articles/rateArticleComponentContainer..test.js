/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { RateArticleComponentContainer } from '../../../components/articles/rating/RateArticleComponentContainer';

describe('RateArticleComponentContainer Component', () => {
  let wrapper;

  const storeItems = {};
  const mockStore = configureStore();
  const store = mockStore(storeItems);

  const props = {
    articleSlug: 'sprinters-hello-git-15',
    rateArticle: jest.fn(),
    updateArticle: jest.fn(),
    e: { preventDefault: jest.fn() },
    ratingsReducer: {
      isUpdatingRating: true,
    },
  };

  beforeEach(() => {
    wrapper = shallow(<RateArticleComponentContainer {...props} />);
    jest.resetAllMocks();
  });

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const wrapperHere = shallow(<RateArticleComponentContainer store={store} {...props} />);
    expect(toJson(wrapperHere)).toMatchSnapshot();
  });

  it('should call handleRating function', async () => {
    const instance = wrapper.instance();
    instance.setState({
      articleSlug: 'sprinters-hello-git-15',
    });
    instance.handleRating(5);
    expect(instance.state.isRating).toBe(false);
    expect(instance.state.newRating).toBe(0);

    sessionStorage.setItem('isLoggedIn', true);

    await instance.handleRating(5);
    expect(props.updateArticle).toBeCalledTimes(1);

    wrapper.setProps({
      ratingsReducer: {
        isUpdatingRating: false,
      },
    });
    await instance.handleRating(5);

    expect(props.updateArticle).toBeCalledTimes(1);
  });


  it('should change state when mouse hovered out star', () => {
    wrapper.state().isRating = true;
    wrapper.instance().handleMouseOut();
    expect(wrapper.state().isRating).toBe(false);
  });

  it('should change state when mouse hovered over star', () => {
    expect(wrapper.state().newRating).toBe(0);
    wrapper.instance().handleMouseOver(3);
    expect(wrapper.state().newRating).toBe(3);
  });
});
