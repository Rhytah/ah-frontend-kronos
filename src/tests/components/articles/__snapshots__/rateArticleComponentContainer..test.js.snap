// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateArticleComponentContainer Component matches snapshot 1`] = `
<ViewArticleRating
  handleMouseOut={[Function]}
  handleMouseOver={[Function]}
  handleRating={[Function]}
  rating={0}
/>
`;

exports[`RateArticleComponentContainer Component should render correctly 1`] = `
<ViewArticleRating
  handleMouseOut={[Function]}
  handleMouseOver={[Function]}
  handleRating={[Function]}
  rating={0}
/>
`;
